File: src/test/java/com/baulsupp/oksocial/completion/UrlCompleterTest.java
package com.baulsupp.oksocial.completion;

import com.baulsupp.oksocial.authenticator.AuthInterceptor;
import com.baulsupp.oksocial.credentials.CredentialsStore;
import com.baulsupp.oksocial.services.test.TestAuthInterceptor;
import com.google.common.collect.Lists;
import java.io.IOException;
import java.util.List;
import okhttp3.OkHttpClient;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class UrlCompleterTest {
  private List<AuthInterceptor<?>> services = Lists.newArrayList(new TestAuthInterceptor());

  private UrlCompleter completer =
      new UrlCompleter(services, new OkHttpClient(), CredentialsStore.NONE,
          CompletionVariableCache.NONE);

  @Test public void returnsAllUrls() throws IOException {
    assertEquals(
        new UrlList(UrlList.Match.HOSTS,
            Lists.newArrayList("https://test.com", "https://test.com/", "https://api1.test.com",
                "https://api1.test.com/")),
        completer.urlList(""));
  }

  @Test public void returnsMatchingUrls() throws IOException {
    assertEquals(
        Lists.newArrayList("https://api1.test.com", "https://api1.test.com/"),
        completer.urlList("https://api1").getUrls("https://api1"));
    assertEquals(Lists.newArrayList(), completer.urlList("https://api2").getUrls("https://api2"));
  }

  @Test public void returnsMatchingEndpointUrls() throws IOException {
    assertEquals(Lists.newArrayList("https://api1.test.com/users.json",
        "https://api1.test.com/usersList.json"),
        completer.urlList("https://api1.test.com/u").getUrls("https://api1.test.com/u"));
  }
}

File: src/test/kotlin/com/baulsupp/oksocial/completion/UrlCompleterTest.kt
package com.baulsupp.oksocial.completion

import com.baulsupp.oksocial.DefaultToken
import com.baulsupp.oksocial.Main
import com.baulsupp.oksocial.authenticator.AuthenticatingInterceptor
import com.baulsupp.oksocial.services.test.TestAuthInterceptor
import kotlinx.coroutines.experimental.runBlocking
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals

class UrlCompleterTest {
  private val main = Main()

  init {
    main.authenticatingInterceptor = AuthenticatingInterceptor(main, listOf(TestAuthInterceptor()))
    main.initialise()
  }

  private val completer = UrlCompleter(main)

  @Test
  fun returnsAllUrls() {
    runBlocking {
      assertEquals(
              UrlList(UrlList.Match.HOSTS,
                      listOf("https://test.com", "https://test.com/",
                              "https://api1.test.com",
                              "https://api1.test.com/")),
              completer.urlList("", DefaultToken))
    }
  }

  @Test
  fun returnsMatchingUrls() {
    runBlocking {
      assertEquals(
              listOf("https://api1.test.com", "https://api1.test.com/"),
              completer.urlList("https://api1", DefaultToken).getUrls("https://api1"))
      assertEquals(listOf(),
              completer.urlList("https://api2", DefaultToken).getUrls("https://api2"))
    }
  }

  @Test
  fun returnsMatchingEndpointUrls() {
    runBlocking {
      assertEquals(listOf("https://api1.test.com/users.json",
              "https://api1.test.com/usersList.json"),
              completer.urlList("https://api1.test.com/u", DefaultToken).getUrls("https://api1.test.com/u"))
    }
  }
}
